libyara/modules/dex.c

```diff
@@ -0,0 +1,743 @@
+  char *magic = yr_malloc(magic_size + 1);
+  if (magic == NULL) {
+    error_num = ERROR_INSUFICIENT_MEMORY;
+    return -1;
+  }
+  memcpy(magic, dex_header->magic, magic_size);
+  magic[magic_size] = '\0';
+  set_string(magic, module, "header.magic");
+
+  set_integer(*dex_header->checksum, module, "header.checksum");
+
+  unsigned long signature_size = member_size(DEX_HEADER, signature);
+  char *signature = yr_malloc(signature_size + 1);
+  if (signature == NULL) {
+    error_num = ERROR_INSUFICIENT_MEMORY;
+    return -1;
+  }
+  memcpy(signature, dex_header->signature, signature_size);
+  signature[signature_size] = '\0';
+  set_string(signature, module, "header.signature");
+
+  set_integer(*dex_header->file_size, module, "header.file_size");
+  set_integer(*dex_header->header_size, module, "header.header_size");
+  set_integer(*dex_header->endian_tag, module, "header.endian_tag");
+  set_integer(*dex_header->link_size, module, "header.link_size");
+  set_integer(*dex_header->link_offset, module, "header.link_offset");
+  set_integer(*dex_header->map_offset, module, "header.map_offset");
+  set_integer(*dex_header->string_ids_size, module, "header.string_ids_size");
+  set_integer(*dex_header->string_ids_offset, module, "header.string_ids_offset");
+  set_integer(*dex_header->type_ids_size, module, "header.type_ids_size");
+  set_integer(*dex_header->type_ids_offset, module, "header.type_ids_offset");
+  set_integer(*dex_header->proto_ids_size, module, "header.proto_ids_size");
+  set_integer(*dex_header->proto_ids_offset, module, "header.proto_ids_offset");
+  set_integer(*dex_header->field_ids_size, module, "header.field_ids_size");
+  set_integer(*dex_header->field_ids_offset, module, "header.field_ids_offset");
+  set_integer(*dex_header->method_ids_size, module, "header.method_ids_size");
+  set_integer(*dex_header->method_ids_offset, module, "header.method_ids_offset");
+  set_integer(*dex_header->class_defs_size, module, "header.class_defs_size");
+  set_integer(*dex_header->class_defs_offset, module, "header.class_defs_offset");
+  set_integer(*dex_header->data_size, module, "header.data_size");
+  set_integer(*dex_header->data_offset, module, "header.data_offset");
+
+  return 0;
+}
+
+
+static int load_string_ids(PDEX_HEADER dex_header, uint8_t *data, size_t data_size, YR_OBJECT *module) {
+  uint32_t offset = yr_le32toh(*dex_header->string_ids_offset);
+  uint32_t string_ids_size = sizeof(STRING_ID_ITEM[*dex_header->string_ids_size]);
+  if (offset + string_ids_size > data_size) {
+    set_integer(1, module, "invalid_dex");
+    return 0;
+  }
+
+  STRING_ID_ITEM *string_ids = yr_malloc(string_ids_size);
+  if (string_ids == NULL) {
+    error_num = ERROR_INSUFICIENT_MEMORY;
+    return -1;
+  }
+  memcpy(string_ids, data + offset, string_ids_size);
+
+  string_ids_size = *dex_header->string_ids_size * sizeof(STRING_ID_ITEM);
+  int i;
+  int p;
+  for (i = 0, p = 0; p < string_ids_size; i += 1, p += sizeof(STRING_ID_ITEM)) {
```

Comment:
There's no need to keep both `p` and `i` for controlling this loop. You can use just:

 `i < *dex_header->string_ids_size`